FWCore/ParameterSet/python/Config.py

```diff
@@ -931,19 +1066,38 @@ def __extractPSet(self,pset):
         adaptor = TopLevelPSetAcessorAdaptor(processPSet,self)
         self._insertInto(adaptor, self.psets_())
         self._insertInto(adaptor, self.vpsets_())
-        self._insertManyInto(adaptor, "@all_modules", all_modules, True)
         self._insertOneInto(adaptor,  "@all_sources", self.source_(), True)
         self._insertOneInto(adaptor,  "@all_loopers", self.looper_(), True)
         self._insertSubProcessesInto(adaptor, "@all_subprocesses", self.subProcesses_(), False)
-        self._insertManyInto(adaptor, "@all_esmodules", self.es_producers_(), True)
-        self._insertManyInto(adaptor, "@all_essources", self.es_sources_(), True)
         self._insertManyInto(adaptor, "@all_esprefers", self.es_prefers_(), True)
         self._insertManyInto(adaptor, "@all_aliases", self.aliases_(), True)
-        self._insertPaths(adaptor)
+        # This will visit all the paths and endpaths that are scheduled to run,
+        # as well as the Tasks associated to them and the schedule. It remembers
+        # the modules, ESSources, ESProducers, and services it visits.
+        nodeVisitor = NodeVisitor()
+        self._insertPaths(adaptor, nodeVisitor)
+        all_modules_onTasksOrScheduled = { key:value for key, value in all_modules.iteritems() if value in nodeVisitor.modules }
+        self._insertManyInto(adaptor, "@all_modules", all_modules_onTasksOrScheduled, True)
+        # Same as nodeVisitor except this one visits all the Tasks attached
+        # to the process.
+        processNodeVisitor = NodeVisitor()
+        for pTask in self.tasks.itervalues():
+            pTask.visit(processNodeVisitor)
+        esProducersToEnable = {}
+        for esProducerName, esProducer in self.es_producers_().iteritems():
+            if esProducer in nodeVisitor.esProducers or not esProducer in processNodeVisitor.esProducers:
```

Comment:
I will add parenthesis here and in the 2 other similar lines. Although it does not affect how it executes, I agree it makes the code easier to read.